flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/ZooKeeperCompletedCheckpointStore.java

```diff
@@ -163,22 +162,50 @@ public void recover() throws Exception {
 
```

Comment:
In line 156, I found code that considers concurrent modifications in the `ZookeeperStateHandleStore`. Just for discussion, I wonder what happens with concurrent modifications after we retrieved `initialCheckpoints`. Couldn't this mean that after we became `initialCheckpoints` (maybe of size 1), this becomes modified and the only handle is suddenly invalid and we can no longer recover? Maybe the code was just written in an overprotective way and this is a non-issue?